src/hardware/serial_terminal.rs

```diff
@@ -25,28 +85,128 @@ impl Write for OutputBuffer {
+    _args: &[&str],
+    context: &mut Context,
+) {
+    writeln!(context, "Available properties:").unwrap();
+    for path in Settings::iter_paths::<heapless::String<32>>("/") {
+        let path = path.unwrap();
+        writeln!(context, "* {path}").unwrap();
+    }
+}
+
+fn handle_reset(
+    _menu: &menu::Menu<Context>,
+    _item: &menu::Item<Context>,
+    _args: &[&str],
+    _context: &mut Context,
+) {
+    cortex_m::peripheral::SCB::sys_reset();
+}
+
+fn handle_property_read(
+    _menu: &menu::Menu<Context>,
+    item: &menu::Item<Context>,
+    args: &[&str],
+    context: &mut Context,
+) {
+    let mut buf = [0u8; 256];
+    if let Some(path) = menu::argument_finder(item, args, "property").unwrap() {
+        let len = match context.flash.settings.get_json(path, &mut buf) {
+            Err(e) => {
+                writeln!(context, "Failed to read {path}: {e}").unwrap();
+                return;
+            }
+            Ok(len) => len,
+        };
+
+        let stringified = core::str::from_utf8(&buf[..len]).unwrap();
+        writeln!(context, "{path}: {stringified}").unwrap();
+    };
+
+    for path in Settings::iter_paths::<heapless::String<32>>("/") {
+        let path = path.unwrap();
+        let len = context.flash.settings.get_json(&path, &mut buf).unwrap();
+        let stringified = core::str::from_utf8(&buf[..len]).unwrap();
+        writeln!(context, "{path}: {stringified}").unwrap();
+    }
+}
+
+fn handle_property_write(
+    _menu: &menu::Menu<Context>,
+    item: &menu::Item<Context>,
+    args: &[&str],
+    context: &mut Context,
+) {
+    let property = menu::argument_finder(item, args, "property")
+        .unwrap()
+        .unwrap();
+    let value = menu::argument_finder(item, args, "value").unwrap().unwrap();
+
+    // Now, write the new value into memory.
+    // TODO: Validate it first?
+    match context.flash.settings.set_json(property, value.as_bytes()) {
+        Ok(_) => {
+            context.flash.save();
+            writeln!(
+                context,
+                "Settings in memory may differ from currently operating settings. \
+        Reset device to apply settings."
+            )
+    .unwrap();
+        }
+        Err(e) => {
+            writeln!(context, "Failed to update {property}: {e:?}").unwrap();
+        }
+    }
+}
+
 pub struct SerialTerminal {
     usb_device: usb_device::device::UsbDevice<'static, UsbBus>,
     usb_serial: usbd_serial::SerialPort<'static, UsbBus>,
+    menu: menu::Runner<'static, Context>,
     output: bbqueue::Consumer<'static, 512>,
-    buffer: OutputBuffer,
 }
 
 impl SerialTerminal {
     pub fn new(
         usb_device: usb_device::device::UsbDevice<'static, UsbBus>,
         usb_serial: usbd_serial::SerialPort<'static, UsbBus>,
+        flash: FlashSettings,
     ) -> Self {
         let (producer, consumer) = OUTPUT_BUFFER.try_split().unwrap();
 
+        let input_buffer =
+            cortex_m::singleton!(: [u8; 255] = [0; 255]).unwrap();
```

Comment:
Why 255 (not 256)?
Is this the same 255/6 as the other one in the property read fn?

Reply:
Fixed in 6182e1c790d6b1f1f5a8e824419c6f877bdcd4f1